Bluetooth: df: Add vendor specific IQ sample reports providing 16 bits integer IQ samples

ppryga-nordic opened the pull request:
Add vendor specific IQ samples report that holds IQ data in 16 bits
signed integer format. Thanks to that we preserve complete accuracy of
IQ samples provided by Nordic Direction Finding Extension in Radio
peripheral. That helps to maintain better accuracy of evaluated
angles with use of reported IQ samples.

Bluetooth Controller has a vendor specific extensions that allows it
to send IQ report events with IQ samples that are 8 bits or 16 bits
signed integer. To use that functionality, there is added common
handler of vendor specific events.

Vendor specific events handling is prioritized to be done by user
provided event handler. If that is not available, then Host generic
implementation enters.

Added vendor specific events that are handled by common Host code
are BT_HCI_EVT_VS_LE_CONNECTIONLESS_IQ_REPORT and BT_HCI_EVT_VS_LE-
_CONNECTION_IQ_REPORT.

The only difference between regular IQ report events is size of
IQ samples, hence implementation of IQ report events is changed to
use the same user callback. To avoid differentiation of user callbacks
new member sample_type was added to bt_df_per_adv_sync_iq_samples-
_report. Also sample member is changed to be a union, to allow easy
access to IQ samples without type casting.

Change implementation of cte_recv_cb to print information about IQ
samples type. The changed code is ready for use of 8 bits and 16 bits
signed integer IQ samples. User has sample level option to enable
printing IQ samples.

ppryga-nordic committed:
Bluetooth: hci: Add VS IQ samples report using 16 bit sign integer sampl

Add vendor specific IQ samples report that holds IQ data in 16 bits
signed integer format. Thanks to that we preserve complete accuracy of
IQ samples provided by Nordic Direction Finding Extension in Radio
peripheral. That helps to maintain better accuracy of evaluated
angles with use of reported IQ samples.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Host: Add common handling of VS IQ report

Bluetooth Controller has a vendor specific extensions that allows it
to send IQ report events with IQ samples that are 8 bits or 16 bits
signed integer. To use that functionality, there is added common
handler of vendor specific events.

Vendor specific events handling is prioritized to be done by user
provided event handler. If that is not available, then Host generic
implementation enters.

Added vendor specific events that are handled by common Host code
are BT_HCI_EVT_VS_LE_CONNECTIONLESS_IQ_REPORT and BT_HCI_EVT_VS_LE-
_CONNECTION_IQ_REPORT.

The only difference between regular IQ report events is size of
IQ samples, hence implementation of IQ report events is changed to
use the same user callback. To avoid differentiation of user callbacks
new member sample_type was added to bt_df_per_adv_sync_iq_samples-
_report. Also sample member is changed to be a union, to allow easy
access to IQ samples without type casting.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
samples: Bluetooth: df: Add use of 16 bits wide IQ sampels

Change implementation of cte_recv_cb to print information about IQ
samples type. The changed code is ready for use of 8 bits and 16 bits
signed integer IQ samples. User has sample level option to enable
printing IQ samples.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic commented:
Rebased onto latest main branch.

jori-nordic commented on doc/connectivity/bluetooth/api/hci.txt:
```diff
@@ -88,6 +88,8 @@ the controller. By default no vendor events are reported.
 	|       | 1    | Fatal Error                                |
 	|       | 2    | Trace Information Event                    |
 	|       | 3    | Scan Request Received Event                |
+	|       | 4    | Connectionless IQ Report Event             |
```
could you increment the rest of the bits?

ppryga-nordic replied:
I do not get, which bits should I change?

jori-nordic replied:
the next ones in the table still have 4, 5 in rfu

jori-nordic commented on include/zephyr/bluetooth/hci_vs.h:
```diff
@@ -263,11 +263,53 @@ struct bt_hci_evt_vs_scan_req_rx {
 #define BT_EVT_MASK_VS_FATAL_ERROR             BT_EVT_BIT(1)
 #define BT_EVT_MASK_VS_TRACE_INFO              BT_EVT_BIT(2)
 #define BT_EVT_MASK_VS_SCAN_REQ_RX             BT_EVT_BIT(3)
+#define BT_EVT_MASK_VS_LE_CONNECTIONLESS_IQ_REPORT BT_EVT_BIT(4)
```
nit: do we want to align the previous definitions?

ppryga-nordic replied:
I didn't align all `#defines` to the long one, to keep them aligned with other `#defines` in the file.
I can change it if you wish.

jori-nordic commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -702,6 +702,26 @@ config BT_DF_CTE_TX_AOD
 	  Enable support for antenna switching during CTE transmission.
 	  Also known as Angle of Departure mode.
 
+config BT_DF_VS_CL_IQ_REPORT_16_BITS_IQ_SAMPLES
```
that's a long name, could `BT_DF_VS_CL_IQ_REPORT_16_BITS` be enough? given we have a help text and all.

ppryga-nordic replied:
Actually I prefer long name instead of forcing every one to look into the definition to be sure what means this `16_BITS` in the option name.

jori-nordic commented on subsys/bluetooth/host/conn.c:
```diff
@@ -3067,6 +3078,18 @@ void bt_hci_le_df_connection_iq_report(struct net_buf *buf)
+#if defined(CONFIG_BT_DF_VS_CONN_IQ_REPORT_16_BITS_IQ_SAMPLES)
+void bt_hci_le_vs_df_connection_iq_report(struct net_buf *buf)
+{
+	bt_hci_le_df_connection_iq_report_common(BT_HCI_EVT_VS_LE_CONNECTION_IQ_REPORT, buf);
```
I guess you're missing something here: you send the same event in both cases.

ppryga-nordic replied:
No ;-) `BT_HCI_EVT_LE_CONNECTION_IQ_REPORT` != `BT_HCI_EVT_VS_LE_CONNECTION_IQ_REPORT`

jori-nordic replied:
yep misread

jori-nordic commented on subsys/bluetooth/host/conn.c:
```diff
@@ -3039,16 +3039,27 @@ int bt_conn_init(void)
+			BT_ERR("Prepare CTE conn IQ report failed %d", err);
+			return;
+		}
+	} else if (IS_ENABLED(CONFIG_BT_DF_VS_CONN_IQ_REPORT_16_BITS_IQ_SAMPLES)) {
```
are you missing an additional condition in this `else if`?

ppryga-nordic replied:
Nooo... If `event != BT_HCI_EVT_LE_CONNECTION_IQ_REPORT` and `IS_ENABLED(CONFIG_BT_DF_VS_CONN_IQ_REPORT_16_BITS_IQ_SAMPLES)` is true, then this `else if` statement will be executed always. If `IS_ENABLED(CONFIG_BT_DF_VS_CONN_IQ_REPORT_16_BITS_IQ_SAMPLES)` is false, then `else` statement will be executed always. 

Compiler should optimize out the unreachable execution path.

jori-nordic replied:
I meant what emil says explicitly here https://github.com/zephyrproject-rtos/zephyr/pull/49117#discussion_r947657636

jori-nordic commented on subsys/bluetooth/host/scan.c:
```diff
@@ -14,6 +14,8 @@
 #include <zephyr/bluetooth/direction.h>
 #include <zephyr/bluetooth/addr.h>
+#include <zephyr/bluetooth/hci.h>
+#include <zephyr/bluetooth/hci_vs.h>
```
I'm not super happy that we include those now, people might be tempted to use more HCI stuff now that they're here.

ppryga-nordic replied:
This file is a Host internal. Is it an issue that there is direct use of HCI header files?
There is no separation layer that covers HCI interface and its types.

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -4974,6 +4986,168 @@ struct net_buf *hci_vs_err_assert(const char *file, uint32_t line)
+	 * HCI event should inform about it by storing single octet with
+	 * special I_sample and Q_sample data.
+	 */
+	samples_cnt = (!iq_report->sample_count ? 1 : iq_report->sample_count);
```
Below you are doing 
```c
samples_cnt = MAX(1, iq_report->sample_count);
```
instead.

Either way of course works, but might make sense to do it the same way both places :)

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -4974,6 +4986,168 @@ struct net_buf *hci_vs_err_assert(const char *file, uint32_t line)
+	sync_handle = ull_sync_handle_get(sync);
+	per_evt_counter = iq_report->event_counter;
+
+	/* If there are no IQ samples due to insufficient resources
+	 * HCI event should inform about it by storing single octet with
+	 * special I_sample and Q_sample data.
+	 */
+	samples_cnt = (!iq_report->sample_count ? 1 : iq_report->sample_count);
```
This value, in case of error, is only used when creating the event with `vs_event`, right?

ppryga-nordic replied:
Yeah... actually this should be changed. In case the insufficient resources are reported, `samples_count` is set to zero, so there is no need to extend the buffer for not used IQ sample. Also assignment:
```
sep->sample[0].i = BT_HCI_LE_CTE_REPORT_NO_VALID_SAMPLE;
sep->sample[0].q = BT_HCI_LE_CTE_REPORT_NO_VALID_SAMPLE;
```
is not needed.

This assignment should be done in case the `packet_status != BT_HCI_LE_CTE_INSUFFICIENT_RESOURCES` but for some reason `iq_report->sample_count` is zero. 

I'll add separate commit with fix for that.

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -3039,16 +3039,27 @@ int bt_conn_init(void)
+			BT_ERR("Prepare CTE conn IQ report failed %d", err);
+			return;
+		}
+	} else if (IS_ENABLED(CONFIG_BT_DF_VS_CONN_IQ_REPORT_16_BITS_IQ_SAMPLES)) {
```
Should this also check if `event == BT_HCI_EVT_VS_LE_CONNECTION_IQ_REPORT`?

ppryga-nordic replied:
Agree, it will make the if statement safer and more readable.

Thalley commented on subsys/bluetooth/host/scan.c:
```diff
@@ -1102,17 +1104,28 @@ void bt_hci_le_biginfo_adv_report(struct net_buf *buf)
+			BT_ERR("Prepare CTE conn IQ report failed %d", err);
+			return;
+		}
+	} else if (IS_ENABLED(CONFIG_BT_DF_VS_CL_IQ_REPORT_16_BITS_IQ_SAMPLES)) {
```
Should this also check if `event == BT_HCI_EVT_VS_LE_CONNECTIONLESS_IQ_REPORT`?

ppryga-nordic replied:
Agree, it will make the if statement safer and more readable.

ppryga-nordic commented on include/zephyr/bluetooth/direction.h:
```diff
@@ -101,10 +101,19 @@ struct bt_df_per_adv_sync_cte_rx_param {
 	const uint8_t *ant_ids;
 };
 
+enum bt_df_iq_samples_type {
+	/** Reported IQ samples have 8 bits signed integer format. Size defined in BT Core 5.3
+	 * Vol 4, Part E sections 7.7.65.21 and 7.7.65.22.
+	 */
+	BT_DF_IQ_SAMPLES_8_BITS_INT,
+	/** Reported IQ samples have 16 bits signed integer format. Vendor specific extension. */
+	BT_DF_IQ_SAMPLES_16_BITS_INT,
```
Agree `_type` is not necessary.

ppryga-nordic commented on include/zephyr/bluetooth/direction.h:
```diff
@@ -118,8 +127,13 @@ struct bt_df_per_adv_sync_iq_samples_report {
 	/** Number of IQ samples in report. */
 	uint8_t sample_count;
+	/** Type of IQ samples provided in a report. */
+	enum bt_df_iq_samples_type sample_type;
```
Could be but I followed `BT 5.3 Core Spec, Vol 4, Part E, section 7.7.65.21` naming convention e.g. `Sample_Count`.

Thalley replied:
Should `bt_df_iq_samples_type` then be `bt_df_iq_sample_type` instead? It's just bugging me that we use plural and singular mixed up like that :D

ppryga-nordic replied:
Indeed it should ;-) I'll change it

ppryga-nordic commented on samples/bluetooth/direction_finding_central/src/main.c:
```diff
@@ -55,6 +57,15 @@ static const char *cte_type2str(uint8_t type)
+static const char *sample_type2str(enum bt_df_iq_samples_type type)
+{
+	switch (type) {
+	case BT_DF_IQ_SAMPLES_8_BITS_INT: return "8 bits int";
+	case BT_DF_IQ_SAMPLES_16_BITS_INT: return "16 bits int";
+	default: return "Unknown";
```
Will refactor the function.

ppryga-nordic committed:
Bluetooth: hci: df: Fix unnecessary IQ sample storage

In case Controller reports to Host insufficient resources to store IQ
samples, according to BT 5.3 Core spec. Vol 4, Part E, sections
7.7.65.21 and 7.7.65.22:
1) packet_status shall be set to BT_HCI_LE_CTE_INSUFFICIENT_RESOURCES
2) sample_count shall be set to zero.
That means there should be no IQ samples stored in IQ report.

HCI implementation stored single IQ sample with value BT_HCI_LE_CTE-
_REPORT_NO_VALID_SAMPLE.

The commit fixes that. No additional memory is added to event buffer
and no IQ sample is stored if sample_count is zero.

Also there is added code that handles situation where ULL provides
IQ report with packet_status indicating valid IQ samples report
but sample_count is zero. In such case Host event stores single IQ
sample with value BT_HCI_LE_CTE_REPORT_NO_VALID_SAMPLE.
In case of vendor specific counter part of the event the value is
BT_HCI_VS_LE_CTE_REPORT_NO_VALID_SAMPLE.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic commented:
- Comments addressed.
- CI issue fixed.
- Added commit to handle unnecessary storage of IQ sample in case of insufficient resource reported by ULL.

Thalley commented on include/zephyr/bluetooth/direction.h:
```diff
@@ -101,10 +101,19 @@ struct bt_df_per_adv_sync_cte_rx_param {
+	/** Reported IQ samples have 8 bits signed integer format. Size defined in BT Core 5.3
+	 * Vol 4, Part E sections 7.7.65.21 and 7.7.65.22.
+	 */
+	BT_DF_IQ_SAMPLES_8_BITS_INT,
+	/** Reported IQ samples have 16 bits signed integer format. Vendor specific extension. */
+	BT_DF_IQ_SAMPLES_16_BITS_INT,
```
```suggestion
	BT_DF_IQ_SAMPLE_8_BITS_INT,
	/** Reported IQ samples have 16 bits signed integer format. Vendor specific extension. */
	BT_DF_IQ_SAMPLE_16_BITS_INT,
```

ppryga-nordic replied:
Ah, I've missed this. I'll change it.

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -2920,11 +2936,15 @@ static void le_df_connectionless_iq_report(struct pdu_data *pdu_rx,
 	 */
-	samples_cnt = (!iq_report->sample_count ? 1 : iq_report->sample_count);
+	if (iq_report->packet_status == BT_HCI_LE_CTE_INSUFFICIENT_RESOURCES) {
+		samples_cnt = 0
```
```suggestion
		samples_cnt = 0;
```

ppryga-nordic replied:
Ups... I should have spot that! Thank you!

Thalley replied:
Did it pass CI before being fixed? Are these being built?

ppryga-nordic replied:
There were other errors in CI builds that could cover this one.
Also the DF tests/code are not always executed by CI. I've got a WPI PR to fix that: https://github.com/zephyrproject-rtos/zephyr/pull/45333

ppryga-nordic commented:
Addressed latest @Thalley change requests.
Added changes to make builds without DF enabled to do not throw warnings.

ppryga-nordic committed:
Bluetooth: Remove default n from common Kconfig

The commit is a minor refactor of the common Host and Controller
Kconfig. There were used "default n" that should be avoided.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic commented:
Added commit that provides minor refactors common Kconfig.

jori-nordic commented on doc/connectivity/bluetooth/api/hci.txt:
```diff
@@ -1325,7 +1325,372 @@ advertising sets, this event shall not be generated.
++-------------------------------+------------+-------------------------------+
+| Event                         | Event Code | Event Parameters              |
++-------------------------------+------------+-------------------------------+
+| LE_Connectionless_IQ_Report   | 0xFF       | Subevent_Code,                |
```
connection iq report

jori-nordic commented on include/zephyr/bluetooth/direction.h:
```diff
@@ -118,8 +127,13 @@ struct bt_df_per_adv_sync_iq_samples_report {
 	uint8_t sample_count;
+	/** Type of IQ samples provided in a report. */
+	enum bt_df_iq_sample sample_type;
 	/** Pinter to IQ samples data. */
```
pointer

jori-nordic commented on include/zephyr/bluetooth/direction.h:
```diff
@@ -165,10 +179,15 @@ struct bt_df_conn_iq_samples_report {
+	enum bt_df_iq_sample sample_type;
 	/** Number of IQ samples in report. */
 	uint8_t sample_count;
 	/** Pinter to IQ samples data. */
```
pointer

jori-nordic commented on subsys/bluetooth/common/Kconfig:
```diff
@@ -214,13 +214,19 @@ config BT_HCI_VS_EXT
-	  Enable emiting HCI Vendor-Specific events for system and Controller error that are
-	  unrecoverable.
+	  Enable emiting HCI Vendor-Specific events for system and Controller
+	  error that are unrecoverable.
```
errors

jori-nordic commented on subsys/bluetooth/common/Kconfig:
```diff
@@ -214,13 +214,19 @@ config BT_HCI_VS_EXT
 	  Host and/or Controller. This enables  Write BD_ADDR, Read Build Info,
 	  Read Static Addresses and Read Key Hierarchy Roots vendor commands.
 
+config BT_HCI_VS_EVT
```
would gating some other kconfig options behind that one make sense? right now it seems it's only used for code.

jori-nordic replied:
or at least add `select`s or `depends on` for the events that depend on that.

ppryga-nordic replied:
This option is added for all possible events that are and may be implemented in future. It is used to guard `vs_events_mask` and `vs_event`. That prevents build time warnings if vendor specific events are not used. Also allows to select only required events to be enabled. 

It is already selected by `BT_CTLR_DF_VS_CL_IQ_REPORT_16_BITS_IQ_SAMPLES`, `BT_CTLR_DF_VS_CONN_IQ_REPORT_16_BITS_IQ_SAMPLES`, `BT_DF_VS_CL_IQ_REPORT_16_BITS_IQ_SAMPLES` and BT_DF_VS_CONN_IQ_REPORT_16_BITS_IQ_SAMPLES`.

jori-nordic replied:
yeah but you have more VS events that that, see the mask you made.

ppryga-nordic replied:
yes there are other events defined in `hci_vs.h` where:
- `bt_hci_evt_vs_fatal_error` is a bit in parallel to other because it uses its private memory for `net_buf`. Needed to be sure that there is a free buf to send event e.g. in case of Hard Fault.
- `bt_hci_evt_vs_trace_info`, `bt_hci_evt_vs_scan_req_rx` are not used by Controller or Host. They are just defined... At least I didn't find any code that uses it.

ppryga-nordic commented:
@jori-nordic thanks for new comments, please re-check.
